src/cstringholder.rs

```diff
@@ -0,0 +1,293 @@
+// Copyright 2018, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+use std::cmp::Ordering;
+use std::ffi::{CStr, CString};
+use std::fmt;
+use std::ops::Deref;
+use std::os::raw::c_char;
+use std::ptr;
+use translate::*;
+use types::{StaticType, Type};
+use libc;
+
+use gobject_ffi;
+use ffi as glib_ffi;
+use value::{FromValueOptional, Value};
+
+#[derive(Hash)]
+pub struct CStringHolder {
+    ptr: *const c_char,
+    length: usize,
+    owned: bool,
+}
+
+impl CStringHolder {
+    pub fn new(ptr: *const c_char) -> Self {
+        assert!(!ptr.is_null());
+        let length = unsafe { libc::strlen(ptr) };
+        Self { ptr, length: length, owned: true }
+    }
+
+    pub fn from(s: &[u8]) -> Self {
+        let slice = match CStr::from_bytes_with_nul(s) {
+            Ok(s) => s,
+            Err(e) => panic!(e),
+        };
+        let ptr = slice.as_ptr();
+        assert!(!ptr.is_null());
+        Self { ptr, length: s.len() - 1, owned: false }
+    }
+
+    pub fn from_string(s: std::string::String) -> Self {
+        let ptr = s.as_ptr();
+        let length = s.len();
+        Self { ptr: ptr as *const c_char, length, owned: false }
```

Comment:
I added this function for cases where a signal handler returns a CStringHolder, like GtkScale::format-value. It's not right indeed :( I wonder if for those cases we should simply don't use CStringHolder.

Reply:
`CStringHolder` IMHO can used in converting C to Rust, otherwise better use `String`, and its `.to_glib_xxx()`,
so better don't use it in signal returns

Reply:
This code is unsafe: `s` will be freed at the end of the function and the pointer stored in `Self` will be pointing at freed memory.